File: cos/olympus/src/main/java/cos/olympus/game/MapUtil.java
package cos.olympus.game;

import cos.ops.Direction;
import org.jetbrains.annotations.Nullable;

import static cos.ops.Direction.*;

public final class MapUtil {

    public static int nextX(Orientable cr) {
        var dir = cr.mv() == null ? cr.sight() : cr.mv();
        return MapUtil.nextX(cr, dir);
    }

    public static int nextY(Orientable cr) {
        var dir = cr.mv() == null ? cr.sight() : cr.mv();
        return MapUtil.nextY(cr, dir);
    }

    public static int nextX(Placeable ort, Direction mv) {
        return switch (mv) {
            case NORTH, SOUTH -> ort.x();
            case WEST -> ort.x() - 1;
            case EAST -> ort.x() + 1;
        };
    }

    public static int nextY(Placeable ort, Direction mv) {
        return switch (mv) {
            case NORTH -> ort.y() - 1;
            case SOUTH -> ort.y() + 1;
            case WEST, EAST -> ort.y();
        };
    }


    public static boolean inZone(Placeable ort, int x, int y, int radius) {
        var oX = ort.x();
        var oY = ort.y();
        return oX <= x + radius && x >= oX - radius && oY <= y + radius && oY >= y - radius;
    }

    public static @Nullable Direction direction(Placeable from, Placeable to) {
        if (from.y() == to.y()) {
            if (from.x() < to.x()) {
                return EAST;
            } else {
                return WEST;
            }
        } else if (from.x() == to.x()) {
            if (from.y() < to.y()) {
                return SOUTH;
            } else {
                return NORTH;
            }
        } else {
            return null;
        }
    }
}

File: cos/src/main/java/cos/olympus/game/MapUtil.kt
package cos.olympus.game

import cos.ops.Direction
import cos.ops.Direction.*

object MapUtil {
    fun nextX(cr: Orientable): Int {
        val dir = if (cr.mv == null) cr.sight else cr.mv!!
        return nextX(cr, dir)
    }

    fun nextY(cr: Orientable): Int {
        val dir = if (cr.mv == null) cr.sight else cr.mv!!
        return nextY(cr, dir)
    }

    fun nextX(ort: Placeable, mv: Direction): Int {
        return when (mv) {
            NORTH, SOUTH -> ort.x
            WEST -> ort.x - 1
            EAST -> ort.x + 1
        }
    }

    fun nextY(ort: Placeable, mv: Direction): Int {
        return when (mv) {
            NORTH -> ort.y - 1
            SOUTH -> ort.y + 1
            WEST, EAST -> ort.y
        }
    }

    fun inZone(ort: Placeable, x: Int, y: Int, radius: Int): Boolean {
        val oX = ort.x
        val oY = ort.y
        return oX <= x + radius && x >= oX - radius && oY <= y + radius && oY >= y - radius
    }

    fun direction(from: Placeable, to: Placeable): Direction? {
        return if (from.y == to.y) {
            if (from.x < to.x) {
                EAST
            } else {
                WEST
            }
        } else if (from.x == to.x) {
            if (from.y < to.y) {
                SOUTH
            } else {
                NORTH
            }
        } else {
            null
        }
    }
}
